akka-stream/src/main/scala/akka/stream/impl/SynchronousFilePublisher.scala

```diff
@@ -0,0 +1,124 @@
+/**
+ * Copyright (C) 2015 Typesafe Inc. <http://www.typesafe.com>
+ */
+package akka.stream.impl
+
+import java.io.{ File, RandomAccessFile }
+import java.nio.channels.FileChannel
+
+import akka.actor.{ DeadLetterSuppression, ActorLogging, Props }
+import akka.io.DirectByteBufferPool
+import akka.stream.actor.ActorPublisherMessage
+import akka.util.ByteString
+
+import scala.concurrent.Promise
+
+/** INTERNAL API */
+private[akka] object SynchronousFilePublisher {
+  def props(f: File, completionPromise: Promise[Long], chunkSize: Int, initialBuffer: Int, maxBuffer: Int) = {
+    require(chunkSize > 0, s"chunkSize must be > 0 (was $chunkSize)")
+    require(initialBuffer > 0, s"initialBuffer must be > 0 (was $initialBuffer)")
+    require(maxBuffer >= initialBuffer, s"maxBuffer must be >= initialBuffer (was $maxBuffer)")
+
+    Props(classOf[SynchronousFilePublisher], f, completionPromise, chunkSize, initialBuffer, maxBuffer)
+  }
+
+  private final case object Continue extends DeadLetterSuppression
+}
+
+/** INTERNAL API */
+private[akka] class SynchronousFilePublisher(f: File, bytesReadPromise: Promise[Long], chunkSize: Int, initialBuffer: Int, maxBuffer: Int)
+  extends akka.stream.actor.ActorPublisher[ByteString]
+  with ActorLogging {
+
+  import SynchronousFilePublisher._
+
+  val buffs = new DirectByteBufferPool(chunkSize, maxBuffer)
```

Comment:
A maybe heretic question: what's the reason for using DirectByteBuffers and FileChannels instead of a simple InputStream based solution (which would then work for any InputStream)?

I would expect this solution to be even slower than the InputStream solution because what would be one native call (`InputStream.read`) now becomes at least two: allocate a direct buffer, FileChannel.read and ByteString(buf). You also don't prevent any buffer copies because `ByteString(buf)` will have to copy the bytes from the native heap into the heap in the same way that the implementation of `InputStream.read` does internally.

Direct buffers usually make sense for asynchronous IO operations but as asynchronous file IO isn't supported on the OS-level (at least under Linux, but also under other OS the interaction with the page cache will always be tricky) using `AsynchronousFileChannel` may not be best solution compared to a simpler one that blocks on a thread controlled by the user instead of blocking in the internal thread pool.